purekit/src/main/kotlin/com/virgilsecurity/purekit/pure/Pure.java

```diff
@@ -225,209 +196,250 @@ public AuthResult authenticateUser(String userId, String password, String sessio
+     * successfully. Represents a regular HTTP exception with code and message.
+     * @throws PureException Please, see {@link PureStorage#selectUser(String)} PureException doc.
+     * @throws DecryptionException If decryption failed.
      */
-    public PureGrant createUserGrantAsAdmin(String userId, VirgilPrivateKey bupsk) throws Exception {
-        UserRecord userRecord = this.storage.selectUser(userId);
+    public PureGrant createUserGrantAsAdmin(String userId, VirgilPrivateKey bupsk)
+        throws ProtocolException, ProtocolHttpException, PureException, DecryptionException {
+
+        UserRecord userRecord = storage.selectUser(userId); // FIXME is userId null/empty ok here?
 
-        byte[] usk = this.crypto.decrypt(userRecord.getEncryptedUskBackup(), bupsk);
+        byte[] usk = crypto.decrypt(userRecord.getEncryptedUskBackup(), bupsk);
 
-        VirgilKeyPair upk = this.crypto.importPrivateKey(usk);
+        VirgilKeyPair upk;
+        try {
+            upk = crypto.importPrivateKey(usk);
+        } catch (CryptoException exception) {
+            throw new IllegalStateException("Exception while importing key. This should not"
+                                                + " happen. Please, contact developer.",
+                                            exception);
+        }
 
         return new PureGrant(upk, userId, null, new Date());
     }
 
     /**
-     * Decrypt encrypted PureGrant that was stored on client-side
-     * @param encryptedGrantString encrypted PureGrant obtained from authenticateUser method
-     * @return PureGrant
-     * @throws InvalidProtocolBufferException FIXME
-     * @throws CryptoException FIXME
+     * Decrypt encrypted PureGrant that was stored on client-side.
+     *
+     * @param encryptedGrantString Encrypted PureGrant obtained from authenticateUser method.
+     *
+     * @return PureGrant.
+     *
+     * @throws ProtocolException Thrown if an error from the PHE service has been parsed
+     * successfully.
+     * @throws ProtocolHttpException Thrown if an error from the PHE service has NOT been parsed
+     * successfully. Represents a regular HTTP exception with code and message.
+     * @throws PureException Please, see {@link PureStorage#selectUser(String)} PureException doc.
+     * @throws InvalidProtocolBufferException If provided encryptedGrantString cannot be parsed as
+     * protobuf message.
      */
-    public PureGrant decryptGrantFromUser(String encryptedGrantString) throws Exception {
-        if (encryptedGrantString == null || encryptedGrantString.isEmpty()) {
-            throw new NullPointerException();
-        }
+    public PureGrant decryptGrantFromUser(String encryptedGrantString)
+        throws ProtocolException, ProtocolHttpException, PureException,
+        InvalidProtocolBufferException {
 
-        byte[] encryptedGrantData = Base64.getDecoder().decode(encryptedGrantString);
+        ValidateUtils.checkNullOrEmpty(new HashMap<String, String>() {{
```

Comment:
This is overkill

Reply:
just similar approach as in all other methods, so all checks are the same

Reply:
Creating HashMap on every function call just to check arguments nullability is overkill, slow and probably could not be optimised by compiler.